File: algorithms/trees/red-black-tree/java/RedBlackTree.java
import java.util.ArrayList;
import java.util.List;

public class RedBlackTree {

    private static final boolean RED = true;
    private static final boolean BLACK = false;

    private static int[] keys;
    private static int[] lefts;
    private static int[] rights;
    private static int[] parents;
    private static boolean[] colors;
    private static int size;
    private static int root;

    private static void init(int capacity) {
        keys = new int[capacity];
        lefts = new int[capacity];
        rights = new int[capacity];
        parents = new int[capacity];
        colors = new boolean[capacity];
        size = 0;
        root = -1;
    }

    private static int newNode(int key) {
        int idx = size++;
        keys[idx] = key;
        lefts[idx] = -1;
        rights[idx] = -1;
        parents[idx] = -1;
        colors[idx] = RED;
        return idx;
    }

    private static void rotateLeft(int x) {
        int y = rights[x];
        rights[x] = lefts[y];
        if (lefts[y] != -1) parents[lefts[y]] = x;
        parents[y] = parents[x];
        if (parents[x] == -1) root = y;
        else if (x == lefts[parents[x]]) lefts[parents[x]] = y;
        else rights[parents[x]] = y;
        lefts[y] = x;
        parents[x] = y;
    }

    private static void rotateRight(int x) {
        int y = lefts[x];
        lefts[x] = rights[y];
        if (rights[y] != -1) parents[rights[y]] = x;
        parents[y] = parents[x];
        if (parents[x] == -1) root = y;
        else if (x == rights[parents[x]]) rights[parents[x]] = y;
        else lefts[parents[x]] = y;
        rights[y] = x;
        parents[x] = y;
    }

    private static void fixInsert(int z) {
        while (z != root && colors[parents[z]] == RED) {
            int gp = parents[parents[z]];
            if (parents[z] == lefts[gp]) {
                int y = rights[gp];
                if (y != -1 && colors[y] == RED) {
                    colors[parents[z]] = BLACK;
                    colors[y] = BLACK;
                    colors[gp] = RED;
                    z = gp;
                } else {
                    if (z == rights[parents[z]]) {
                        z = parents[z];
                        rotateLeft(z);
                    }
                    colors[parents[z]] = BLACK;
                    colors[parents[parents[z]]] = RED;
                    rotateRight(parents[parents[z]]);
                }
            } else {
                int y = lefts[gp];
                if (y != -1 && colors[y] == RED) {
                    colors[parents[z]] = BLACK;
                    colors[y] = BLACK;
                    colors[gp] = RED;
                    z = gp;
                } else {
                    if (z == lefts[parents[z]]) {
                        z = parents[z];
                        rotateRight(z);
                    }
                    colors[parents[z]] = BLACK;
                    colors[parents[parents[z]]] = RED;
                    rotateLeft(parents[parents[z]]);
                }
            }
        }
        colors[root] = BLACK;
    }

    private static void insert(int key) {
        int y = -1;
        int x = root;
        while (x != -1) {
            y = x;
            if (key < keys[x]) x = lefts[x];
            else if (key > keys[x]) x = rights[x];
            else return;
        }
        int node = newNode(key);
        parents[node] = y;
        if (y == -1) root = node;
        else if (key < keys[y]) lefts[y] = node;
        else rights[y] = node;
        fixInsert(node);
    }

    private static void inorder(int node, List<Integer> result) {
        if (node == -1) return;
        inorder(lefts[node], result);
        result.add(keys[node]);
        inorder(rights[node], result);
    }

    public static int[] rbInsertInorder(int[] arr) {
        init(arr.length + 1);
        for (int val : arr) insert(val);
        List<Integer> result = new ArrayList<>();
        inorder(root, result);
        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}

File: algorithms/trees/red-black-tree/c/red_black_tree.c
#include "red_black_tree.h"
#include <stdlib.h>

#define RB_RED 1
#define RB_BLACK 0

typedef struct RBNode {
    int key;
    struct RBNode* left;
    struct RBNode* right;
    struct RBNode* parent;
    int color;
} RBNode;

static RBNode* root_g;

static RBNode* create_node(int key) {
    RBNode* node = (RBNode*)malloc(sizeof(RBNode));
    node->key = key;
    node->left = NULL;
    node->right = NULL;
    node->parent = NULL;
    node->color = RB_RED;
    return node;
}

static void rotate_left(RBNode* x) {
    RBNode* y = x->right;
    x->right = y->left;
    if (y->left) y->left->parent = x;
    y->parent = x->parent;
    if (!x->parent) root_g = y;
    else if (x == x->parent->left) x->parent->left = y;
    else x->parent->right = y;
    y->left = x;
    x->parent = y;
}

static void rotate_right(RBNode* x) {
    RBNode* y = x->left;
    x->left = y->right;
    if (y->right) y->right->parent = x;
    y->parent = x->parent;
    if (!x->parent) root_g = y;
    else if (x == x->parent->right) x->parent->right = y;
    else x->parent->left = y;
    y->right = x;
    x->parent = y;
}

static void fix_insert(RBNode* z) {
    while (z->parent && z->parent->color == RB_RED) {
        RBNode* gp = z->parent->parent;
        if (z->parent == gp->left) {
            RBNode* y = gp->right;
            if (y && y->color == RB_RED) {
                z->parent->color = RB_BLACK;
                y->color = RB_BLACK;
                gp->color = RB_RED;
                z = gp;
            } else {
                if (z == z->parent->right) {
                    z = z->parent;
                    rotate_left(z);
                }
                z->parent->color = RB_BLACK;
                z->parent->parent->color = RB_RED;
                rotate_right(z->parent->parent);
            }
        } else {
            RBNode* y = gp->left;
            if (y && y->color == RB_RED) {
                z->parent->color = RB_BLACK;
                y->color = RB_BLACK;
                gp->color = RB_RED;
                z = gp;
            } else {
                if (z == z->parent->left) {
                    z = z->parent;
                    rotate_right(z);
                }
                z->parent->color = RB_BLACK;
                z->parent->parent->color = RB_RED;
                rotate_left(z->parent->parent);
            }
        }
    }
    root_g->color = RB_BLACK;
}

static void insert_key(int key) {
    RBNode* y = NULL;
    RBNode* x = root_g;
    while (x) {
        y = x;
        if (key < x->key) x = x->left;
        else if (key > x->key) x = x->right;
        else return;
    }
    RBNode* node = create_node(key);
    node->parent = y;
    if (!y) root_g = node;
    else if (key < y->key) y->left = node;
    else y->right = node;
    fix_insert(node);
}

static void inorder(RBNode* node, int* result, int* idx) {
    if (!node) return;
    inorder(node->left, result, idx);
    result[(*idx)++] = node->key;
    inorder(node->right, result, idx);
}

static void free_tree(RBNode* node) {
    if (!node) return;
    free_tree(node->left);
    free_tree(node->right);
    free(node);
}

void rb_insert_inorder(const int* arr, int n, int* result, int* result_size) {
    root_g = NULL;
    for (int i = 0; i < n; i++) insert_key(arr[i]);
    *result_size = 0;
    inorder(root_g, result, result_size);
    free_tree(root_g);
    root_g = NULL;
}
